### PR TITLE
tpu composer support

### PR DESCRIPTION
**Pull request by florescl**
some throughput/profiling: https://wandb.ai/mosaic-ml/tpu-r50?workspace=user-laura

### CODE REVIEW

File: composer/trainer/devices/__init__.py [hunk 1]
```diff
@@ -8,8 +8,10 @@
 from composer.trainer.devices.device import Device as Device
 from composer.trainer.devices.device_cpu import DeviceCPU as DeviceCPU
 from composer.trainer.devices.device_gpu import DeviceGPU as DeviceGPU
+from composer.trainer.devices.device_gpu import DeviceGPU as DeviceTPU
```
Comment (ravi-mosaicml) [inline composer/trainer/devices/__init__.py:11]:
```suggestion
from composer.trainer.devices.device_tpu import DeviceTPU as DeviceTPU
```

File: composer/trainer/devices/device_tpu.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+import logging
+from contextlib import contextmanager
+from typing import Generator, TypeVar, Union
+
+import torch
+
+from composer.core.precision import Precision
+from composer.trainer.devices.device import Device, T_nnModule
+import torch_xla
```
Comment (ravi-mosaicml) [inline composer/trainer/devices/device_tpu.py:15]: Can this be a conditional dependency? See https://github.com/mosaicml/composer/blob/dev/STYLE_GUIDE.md#61-external-dependencies

File: composer/trainer/devices/device_tpu.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+    def __init__(self):
+        import torch_xla.core.xla_model as xm
+
+        self._device = xm.xla_device()
+        
+    def module_to_device(self, module: T_nnModule) -> T_nnModule:
+        import torch_xla.distributed.xla_multiprocessing as xmp
+        
+        WRAPPED_MODEL = xmp.MpModelWrapper(module)
```
Comment (ravi-mosaicml) [inline composer/trainer/devices/device_tpu.py:38]:
```suggestion
        warpped_model = xmp.MpModelWrapper(module)
```

File: composer/trainer/devices/device_tpu.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    This class takes no arguments.
+    """
+
+    def __init__(self):
+        import torch_xla.core.xla_model as xm
+
+        self._device = xm.xla_device()
+        
+    def module_to_device(self, module: T_nnModule) -> T_nnModule:
+        import torch_xla.distributed.xla_multiprocessing as xmp
```
Comment (ravi-mosaicml) [inline composer/trainer/devices/device_tpu.py:36]: Two questions:

1. Does this assume that the module is wrapped with `DDP`? If so, can we relax this requirement?
2. Does this work with deepspeed?

File: composer/trainer/devices/device_tpu.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        
+        WRAPPED_MODEL = xmp.MpModelWrapper(module)
+        return WRAPPED_MODEL.to(self._device)
+
+    def tensor_to_device(self, tensor: torch.Tensor) -> torch.Tensor:
+        return tensor.to(self._device)
+
+    @contextmanager
+    def precision_context(self, precision: Union[str, Precision]) -> Generator[None, None, None]:
```
Comment (ravi-mosaicml) [inline composer/trainer/devices/device_tpu.py:45]: The `precision_context` was removed; see https://github.com/mosaicml/composer/blob/dev/composer/core/precision.py#L35

File: composer/trainer/devices/device_tpu.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+            yield
+        else:
+            raise ValueError(f"Precision {precision} not supported for a tpu")
+
+    def state_dict(self) -> StateDict:
+        return {}
+
+    def load_state_dict(self, state: StateDict) -> None:
+        if len(state) != 0:
+            raise ValueError("CPU device has no state.")
```
Comment (ravi-mosaicml) [inline composer/trainer/devices/device_tpu.py:57]:
```suggestion
            raise ValueError("TPU device has no state.")
```

File: composer/trainer/trainer.py [hunk 7]
```diff
@@ -666,7 +668,8 @@ def __init__(
         if self.deepspeed_enabled or dist.get_world_size() > 1:
             # deepspeed requires torch.distributed to be initialized, even if the world size is 1
             # distributed is always required with multi-rank training
-            dist.initialize_dist(self._device.dist_backend, datetime.timedelta(seconds=dist_timeout))
+            if self._device is not "tpu":
+                dist.initialize_dist(self._device.dist_backend, datetime.timedelta(seconds=dist_timeout))
```
Comment (ravi-mosaicml) [inline composer/trainer/trainer.py:672]: Do TPUs support distributed training?

File: composer/trainer/trainer.py [hunk 8]
```diff
@@ -1023,7 +1026,14 @@ def saved_checkpoints(self) -> List[Tuple[Timestamp, List[pathlib.Path]]]:
     def fit(self):
         """Train and evaluate the model on the provided data."""
         # Print any exception, so it can be caputred by any callbacks or loggers (e.g. WandB, FileLogger)
-        self._train_loop()
+
+        if self._device == "tpu":
+            import torch_xla.distributed.xla_multiprocessing as xmp            
+            def _mp_fn(index):
+                self._train_loop()
+            xmp.spawn(_mp_fn, args=(), nprocs=8,start_method='fork')
```
Comment (ravi-mosaicml) [inline composer/trainer/trainer.py:1034]: I'm confused here -- can you elaborate on this change? For distributed training, we use a launch script -- see https://docs.mosaicml.com/en/stable/trainer/distributed_training.html

File: composer/trainer/trainer.py [hunk 9]
```diff
@@ -1358,7 +1380,7 @@ def _train_microbatches(self, microbatches: Sequence[Batch], ddp_sync: bool = Tr
                 is_final_microbatch = microbatch_idx + 1 == len(microbatches)
                 self._train_microbatch(use_grad_scaling, current_batch_size, total_loss, is_final_microbatch)
 
-            # Unscale gradients before `Event.AFTER_TRAIN_BATCH`
+            # Unscale gradients before `Event.AFTER_TRAIN_BATCH`
```
Comment (ravi-mosaicml) [inline composer/trainer/trainer.py:1383]:
```suggestion
            # Unscale gradients before `Event.AFTER_TRAIN_BATCH`
```

File: composer/trainer/trainer_hparams.py [hunk 10]
```diff
@@ -498,7 +499,7 @@ def initialize_object(self) -> Trainer:
 
         # initialize distributed early so that it's already initialized when dataloders
         # are created.
-        if dist.get_world_size() > 1:
+        if dist.get_world_size() > 1 and device is not "tpu":
```
Comment (ravi-mosaicml) [inline composer/trainer/trainer_hparams.py:502]: Same comment re: distributed training

File: composer/utils/module_surgery.py [hunk 11]
```diff
@@ -392,8 +392,12 @@ def replace_params_in_optimizer(old_params: Iterable[torch.nn.parameter.Paramete
         if old_params is None or new_params is None:
             raise RuntimeError("old_params and new_params have different lengths.")
 
+        '''
```
Comment (ravi-mosaicml) [inline composer/utils/module_surgery.py:395]: I presume this change should be uncommented?

File: setup.py [hunk 12]
```diff
@@ -64,8 +64,8 @@ def package_files(prefix: str, directory: str, extension: str):
     "tqdm>=4.62.3,<5",
     "torchmetrics>=0.7.0,<0.8",
     "torch_optimizer>=0.1.0,<0.2",
-    "torchvision>=0.10.0",  # torchvision has strict pytorch requirements
-    "torch>=1.9,<2",
+    #"torchvision>=0.10.0",  # torchvision has strict pytorch requirements
```
Comment (ravi-mosaicml) [inline setup.py:67]: Can these changes be reverted?

File: composer/yamls/models/resnet50_synthetic.yaml [hunk 13]
```diff
@@ -32,8 +32,8 @@ model:
       - bn_uniform
     num_classes: 1000
 max_duration: 90ep
-train_batch_size: 2048
-eval_batch_size: 2048
+train_batch_size: 1024
```
Comment (ravi-mosaicml) [inline composer/yamls/models/resnet50_synthetic.yaml:35]: Can these changes be reverted?

File: composer/trainer/trainer.py [hunk 14]
```diff
@@ -1285,15 +1297,25 @@ def _train_batch(self, use_grad_scaling: bool):
...
+                                total_loss = xm.optimizer_step(optimizer)
+                            else:
+                                total_loss = optimizer.step(
+                                    closure=lambda **kwargs: self._train_microbatches(microbatches, **kwargs).item())
+                              
                 else:
                     total_loss = self._train_microbatches(microbatches)
                     for optimizer in self.state.optimizers:
                         if use_grad_scaling:
+                            xm.optimizer_step(optimizer)
```
Comment (ravi-mosaicml) [inline composer/trainer/trainer.py:1310]: Should this be conditional on whether the device is a TPU?

File: composer/trainer/trainer.py [hunk 15]
```diff
@@ -1285,15 +1297,25 @@ def _train_batch(self, use_grad_scaling: bool):
                             total_loss = self.state.scaler.step(
                                 optimizer, closure=lambda **kwargs: self._train_microbatches(microbatches, **kwargs))
                         else:
-                            total_loss = optimizer.step(
-                                closure=lambda **kwargs: self._train_microbatches(microbatches, **kwargs).item())
+                            if self._device == "tpu":
+                                total_loss = xm.optimizer_step(optimizer)
```
Comment (ravi-mosaicml) [inline composer/trainer/trainer.py:1301]: Do TPUs support closures? If not, can that raise an error?

**Review by ravi-mosaicml — COMMENTED** [review]
Good first draft!

In addition to the specific comments below:

1. Can you also add a test for TPUs? For the test, can you mark the test with `@pytest.mark.tpu` (I'll update the pytest / Jenkins config to support this marker). In addition, can you include detailed instructions for how one would run this test? (Since we likely won't be able to run it automatically via Jenkins, I want to make sure that I and others will know how to run it!) I am thinking that the test matrix should look something like the following:
  
    1. Basic TPU training -- does it run?
    1. Checkpointing tests with TPU -- can a TPU model be checkpointed and resumed? (See the existing `test_checkpoint.py`)
    1. Algorithm tests with TPU -- for each algorithm, apply it and train with TPU. I presume some algorithms will fail -- that's fine, but let's mark them with `@pytest.mark.xfail` so we can track it and fix it later.
2. Can you update the docs with a tutorial and/or notebook on how to train with TPUs? OK if this happens in a future PR.
3. After #948 is merged, can you update the defaults so that the device will default to xla_device if on a TPU instance?

**Comment by florescl** [comment]
closing, will open a fresh one from my fork.

**Pull request closed without merging**